### ISSUE TITLE
PAL mode overriden in sunxi/dip.c

### ISSUE THREAD

**Issue by nazar554** [NextThingCo/CHIP-u-boot#5]
https://bbs.nextthing.co/t/pal-woes-setenv-command-not-found/6291/17

```
static void dip_setup_pocket_display(enum disp_output display)
{
        char *env_var;

        switch (display) {
        case DISPLAY_RGB_HDMI_BRIDGE:
                env_var = "sunxi:1024x768-24@60,monitor=hdmi";
                break;

        case DISPLAY_RGB_VGA_BRIDGE:
                env_var = "sunxi:1024x768-24@60,monitor=vga";
                break;

        case DISPLAY_RGB_POCKET:
                env_var = "sunxi:480x272-16@60,monitor=lcd";
                break;

        default:
                env_var = "sunxi:720x480-24@60,monitor=composite-ntsc,overscan_x=40,overscan_y=20";
                break;
        }

        setenv("video-mode", env_var);
}
```

**Comment by biasedlogic** [comment]
This code apparently breaks any attempt to configure any non-standard display, not only PAL vs NTSC thing.
More: [An Attempt to connect a 800x480 TFT to Chip](https://bbs.nextthing.co/t/pocket-chip-with-800x480-lcd-hardware-works-how-to-configure/8531/9)

**Comment by mripard** [comment]
Sorry for the late answer guys.

I just pushed some changes to our u-boot today (in the nextthing/2016.01/next branch) that should solve both these issues:
- There's a new environment variable called "dip-auto-video" that, if set to "no", will disable the setting of the kernel commandline and video-mode based on the connected DIP. That should solve @biasedlogic's use case
- There's a second environment variable called "tv-mode", that you can set to either "pal" or "ntsc", that will allow the auto-detection code above to adjust both video-mode and the kernel commandline according to whatever is set there. It should fix @nazar554 use case.

**Comment by biasedlogic** [comment]
How about custom mode definition? Can the timing parameters be set without recompiling the uboot?

**Comment by mripard** [comment]
Not yet unfortunately. That would require a signficant rework of u-boot's code. Feel free to add a new issue for that.

**Comment by biasedlogic** [comment]
Ok. 
Where are the mode definitions put now? 
Can u-boot be mix - matched with the rest of the system image? Is there any how to related to building and flashing u boot?

**Comment by mripard** [comment]
So far, in CONFIG_VIDEO_LCD_MODE, in configs/CHIP_defconfig.

However, It might be easier than I originally thought. How experienced are you with compiling U-Boot, using a terminal, etc?

**Comment by biasedlogic** [comment]
Never compiled u-boot, but I have experience with low level coding, embedded systems, firmware, however mainly in the bare metal systems, where the os is part of the application. I feel at home with command line, at least much more than with a fancy one click tool, that never works the way you thought it would ...

**Comment by biasedlogic** [comment]
So, I checked me the sources here, and after a bit of tweaking I could adjust CHIP_defconfig and execute

`make CROSS_COMPILE=arm-linux-gnueabihf- CHIP_config`
`make CROSS_COMPILE=arm-linux-gnueabihf-`

and it seems it has worked. How do I go about installing the result, @mripard ?

**Comment by mripard** [comment]
You did just fine ! :)

You can have a look at NextThingCo/CHIP-Linux#45 for instructions on how to quick test that.

I just pushed an (untested) branch mr/2016.01/video-test, if you want to give it a try.

It adds a new environment variable in u-boot, "video-timings", that you can use to create a whole set of timings.

The syntax is defined here:  https://github.com/NextThingCo/CHIP-u-boot/blob/nextthing/2016.01/next/drivers/video/videomodes.c#L39, and you have an example for the pocket here: https://github.com/NextThingCo/CHIP-u-boot/blob/nextthing/2016.01/next/configs/CHIP_defconfig#L12

Let me know how it goes.

**Comment by biasedlogic** [comment]
Thanks! 
I'm out on a business trip for a couple of days, but i hope to test it over the weekend. I'll keep you posted. 

Marek

**Comment by mripard** [comment]
I just gave this a try, and it seems to work as expected, I merged that in nextthing/2016.01/next

**Comment by biasedlogic** [comment]
sounds great! I'm not yet back, but will test ASAP, which unfortunately may mean late weekend.

**Comment by norgeous** [comment]
@mripard Forgive my stupidity, but in reference to:

> I merged that in nextthing/2016.01/next

Will this be available in version 4.4 on http://flash.getchip.com/ or some other version?
and if so when?

Would love to get my PAL tv working with CHIP!

**Comment by mripard** [comment]
It should be eventually yes, but for an exact time frame, you'll need to ping @zerotri or @computermouth

**Comment by norgeous** [comment]
Hello @zerotri and/or @computermouth, when will the `nextthing/2016.01/next` branch code be available for flashing at http://flash.getchip.com/? [Lots of users](https://bbs.nextthing.co/t/pal-woes-setenv-command-not-found/6291) can't use chip 4.4 with PAL tv's. Sad times.

**Comment by PLJohnny** [comment]
@zerotri @computermouth It's been quite some time but the problem still persists. I even got an USB-UART cable just to learn that the u-boot variable gets overwritten. Are you planning to update that, since it's reportedly already fixed?

**Comment by maqsudbek** [comment]
@zerotri , @computermouth  please update the system to "nextthing/2016.01/next"  branch asap. Because a lot of chip users live in Europe, can not use PAL right now.

**Comment by Bigfellas** [comment]
I have tried all the other posts and cant get Pal to work

**Comment by mohaggag** [comment]
still no pal mode !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

**Comment by esalgado** [comment]
Still no pal mode!

**Comment by esalgado** [comment]
@mripard you mention in your comment from 31/August/2016 

> You can have a look at NextThingCo/CHIP-Linux#45 for instructions on how to quick test that.

but I could not find any documentation at the CHIP-Linux github, even trying "https://github.com/NextThingCo/CHIP-linux#45"

would you be so kind to help me finding out that documentation on how to flash the modified u-boot image? :)

Thanks!!

**Comment by buzztiaan** [comment]
Does PAL work yet? or would i need to install 4.3

**Comment by buzztiaan** [comment]
@esalgado  the link is https://github.com/NextThingCo/CHIP-linux/issues/45

**Comment by buzztiaan** [comment]
fyi, 4.3 does PAL perfectly (dug up a 8GB CHIP)

**Comment by esalgado** [comment]
Yes I confirm it works,  it is my old old Sanyo 14" tv that does strange things and may not be 100% compatible when using the combined RGB cable. With an equally old Philipps TV the PAL mode is very colorful!

**Comment by buzztiaan** [comment]
you confirm the patched uboot works for PAL on 4.4?

**Comment by esalgado** [comment]
I confirm that the UNMODIFIED stock u-boot works fine with pal in one of my TVs 😀 sorry if my writing made it look like it had to do with the patches!

**Comment by buzztiaan** [comment]
thats not true, the UNMODIFIED uboot -only- outputs NTSC, meaning your Philips TV just accepts NTCS

(also , its not a combined RGB cable, its composite)

**Comment by esalgado** [comment]
Hi Buzz, 
ok, my mistake, I believed what I was saying because:
-Raspberry Pi 1 Model B, is shown with color in the Sanyo TV. Only RCA Yellow cable.
-CHIP did not show any color in that TV, no matter what settings I tried, but it did show color in any other TV I had tested it in, with the composite cable.
-Raspberry Pi 3 Model B, did not show any color in that TV, but it did show in any other TV I had tested it in, with the composite cable.

Anyway I sold my CHIP a couple of months ago, tired of the lack of support!
Have fun!!!! :)
